packages/go_router/lib/src/builder.dart

```diff
@@ -172,6 +172,10 @@ class RouteBuilder {
       // The key to provide to the ShellRoute's Navigator.
       final GlobalKey<NavigatorState> shellNavigatorKey = route.navigatorKey;
 
+      // The observers list for the ShellRoute's Navigator.
+      final List<NavigatorObserver> observers =
```

Comment:
I think we need to create a new herocontroller here. It doesn't make sense to ask user to pass in controller if it is needed anyway

Reply:
@chunhtai I get your point, but not all apps have the need of a HeroController as it's only needed if user actually uses a Hero Widget, besides this in general solves the need for a middleware for the nested Navigator. I don't see a reason why a ShellRoute should not expose observers for the nested Navigator.

Reply:
@chunhtai  Another use case, would be for `FirebaseAnalyticsObserver`. Currently, it is only observed for routes at root `GoRouter`, by allowing this `NavigatorObserver` would allow us to track navigation in google analytics.

Reply:
Shall we do that in a new PR and merge this?

Reply:
@chunhtai Shall we do that in a new PR and merge this?

Reply:
I think we should add it in this PR.

Reply:
I would like to clarify why we should do this now. If we merge this as is, people will start to insert their own HeroController so they can make hero to work. It will then be a breaking change if we create the herocontroller by default in the later change

Reply:
@chunhtai, as for creating the default `HeroController`, how could we determine which static factory method to use for `HeroController` creation (`createCupertinoHeroController` or `createMaterialHeroController`) within design agnostic `go_router`?